File: gunhoflash/alien-dictionary.ts
class Letter {
  value: string;
  isUsed: boolean = false;
  before: Set<Letter> = new Set();
  after: Set<Letter> = new Set();

  constructor(value: string) {
    this.value = value;
  }
}

/**
 * convert character to number
 * @param c {string} - character
 * @returns {number} - index of character
 */
function charToInt(c: string) {
  return c.charCodeAt(0) - 97;
}

/**
 * returns first different character between two strings
 * if one string is a prefix of the other, returns null
 * @param a {string} - first word
 * @param b {string} - second word
 * @returns {string[] | null} - first different character between two strings
 */
function getFirstDiffLetters(a: string, b: string) {
  let i = 0;
  while (a[i] === b[i]) i++;
  return a[i] && b[i] ? [a[i], b[i]] : null;
}

function alienOrder(words: string[]): string {
  const letters = new Map<string, Letter>();

  Array(26).fill(0).forEach((_, i) => {
    letters.set(String.fromCharCode(i + 97), new Letter(String.fromCharCode(i + 97)));
  });

  let prevWord = '';
  for (const word of words) {
    word.split('').forEach(c => {
      letters.get(c)!.isUsed = true;
    });

    const diffLetters = getFirstDiffLetters(prevWord, word);
    if (diffLetters) {
      const [ahead, behind] = diffLetters;
      const aheadLetter = letters.get(ahead)!;
      const behindLetter = letters.get(behind)!;
      aheadLetter.after.add(behindLetter);
      behindLetter.before.add(aheadLetter);
    }

    prevWord = word;
  }

  // delete not used letters
  letters.forEach(letter => {
    if (!letter.isUsed) letters.delete(letter.value);
  });

  let answer = '';
  while (letters.size) {
    const lettersWithoutBefore = Array.from(letters.values()).filter(letter => letter.before.size === 0);
    if (lettersWithoutBefore.length === 0) break;
    lettersWithoutBefore
    .sort((a, b) => charToInt(a.value) - charToInt(b.value))
    .forEach(letter => {
      letters.delete(letter.value);
      answer += letter.value;
      letter.after.forEach(l => l.before.delete(letter));
    });
  }

  if (letters.size) return '';
  return answer;
}

File: gunhoflash/alien-dictionary.py
from typing import List

def charToInt(c):
  return ard[c] - 97

def getFirstDiffLetters(a, b):
  index = next((i for i in range(min(len(a), len(b))) if a[i] != b[i]), None)
  if index is not None:
    return [a[index], b[index]]
  return None

class Letter:
  def __init__(self, value):
    self.value = value
    self.isUsed = False
    self.before = set()
    self.after = set()

class Solution:
  def alien_order(self, words: List[str]) -> str:

    letters = dict()

    for i in range(26):
      letters[chr(i + 97)] = Letter(chr(i + 97))

    prevWord = ''
    for word in words:
      for c in word:
        letters.get(c).isUsed = True

      diffLetters = getFirstDiffLetters(prevWord, word)
      if diffLetters:
        [ahead, behind] = diffLetters
        aheadLetter = letters.get(ahead)
        behindLetter = letters.get(behind)
        aheadLetter.after.add(behindLetter)
        behindLetter.before.add(aheadLetter)

      prevWord = word

    # delete not used letters
    letters = dict(filter(lambda pair: pair[1].isUsed, letters.items()))

    answer = ''
    while len(letters):
      lettersWithoutBefore = list(filter(lambda letter: len(letter.before) == 0, letters.values()))
      if len(lettersWithoutBefore) == 0: break
      lettersWithoutBefore = sorted(lettersWithoutBefore, key=lambda letter: letter.value)
      for letter in lettersWithoutBefore:
        letters.pop(letter.value)
        answer += letter.value
        for l in letter.after:
          l.before.remove(letter)

    return '' if len(letters) > 0 else answer
